src/machine_learning/loss_function/marginal_ranking.rs

```diff
@@ -0,0 +1,100 @@
+/// Marginal Ranking
+///
+/// The 'mrg_ranking_loss' function calculates the Marginal Ranking loss, which is a
+/// loss function used for ranking problems in machine learning.
+///
+/// ## Formula
+///
+/// For a pair of values `x_first` and `x_second`, `margin`, and `y_true`,
+/// the Marginal Ranking loss is calculated as:
+///
+///  - loss = `max(0, -y_true * (x_first - x_second) + margin)`.
+///
+/// It returns the average loss by dividing the `total_loss` by total no. of
+/// elements.
+///
+/// Pytorch implementation:
+/// https://pytorch.org/docs/stable/generated/torch.nn.MarginRankingLoss.html
+/// https://gombru.github.io/2019/04/03/ranking_loss/
+/// https://vinija.ai/concepts/loss/#pairwise-ranking-loss
+///
+
+pub fn mrg_ranking_loss(
+    x_first: &[f64],
+    x_second: &[f64],
+    margin: f64,
+    y_true: f64,
+) -> Result<f64, MarginalRankingLossError> {
+    if x_first.len() != x_second.len() {
+        return Err(MarginalRankingLossError::InputsHaveDifferentLength);
+    }
+    if x_first.is_empty() {
+        return Err(MarginalRankingLossError::EmptyInputs);
+    }
+    if margin < 0.0 {
+        return Err(MarginalRankingLossError::InvalidValues);
+    }
+    if y_true != 1.0 && y_true != -1.0 {
+        return Err(MarginalRankingLossError::InvalidValues);
+    }
+
+    let mut total_loss: f64 = 0.0;
+    for (f, s) in x_first.iter().zip(x_second.iter()) {
+        let loss: f64 = (margin - y_true * (f - s)).max(0.0);
+        total_loss += loss;
+    }
+    Ok(total_loss / (x_first.len() as f64))
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum MarginalRankingLossError {
+    InputsHaveDifferentLength,
+    EmptyInputs,
+    InvalidValues,
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    macro_rules! test_with_wrong_inputs {
+        ($($name:ident: $inputs:expr,)*) => {
+            $(
+                #[test]
+                fn $name() {
+                    let (x_first, x_second, margin, y_true, expected) = $inputs;
+                    assert_eq!(mrg_ranking_loss(&x_first, &x_second, margin, y_true), expected);
```

Comment:
The _error handing_ is _symmetric_ in this case, therefore one could use something like:
```suggestion
                    let (vec_a, vec_b, margin, y_true, expected) = $inputs;
                    assert_eq!(mrg_ranking_loss(&vec_a, &vec_b, margin, y_true), expected);
                    assert_eq!(mrg_ranking_loss(&vec_b, &vec_a, margin, y_true), expected);
```

Reply:
Makes sense, will add that to the code.